packages/workshop-app/app/components/toast.tsx

```diff
@@ -0,0 +1,254 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+import { renderToString } from 'react-dom/server'
+import * as ToastPrimitive from '@radix-ui/react-toast'
+import { AnimatePresence, motion } from 'framer-motion'
+import type { DefaultColors } from 'tailwindcss/types/generated/colors'
+import clsx from 'clsx'
+import { useEventListener } from '~/utils/misc'
+import Icon from './icons'
+
+// keys need to match Icon in ./icons, value need to match tailwind color
+const colors: Record<ToastVariant, keyof DefaultColors> = {
+	action: 'black', // this one don't use an icon
+	Error: 'red',
+	Notify: 'blue',
+	Success: 'green',
+}
+
+type ToastProps = Omit<ToastEventProps, 'content'>
+
+function BaseToast({
+	title,
+	children,
+	visible = Boolean(title || children),
+	duration = 7000,
+	autoClose,
+	variant,
+	onDismiss,
+}: ToastProps) {
+	const hiddenAction = variant === 'action' && !visible
+	const [open, setOpen] = useState(variant === 'action' || visible)
+	const toastRef = useRef<HTMLLIElement>(null)
+
+	const isAction = variant === 'action'
+	useEffect(() => {
+		setTimeout(() => {
+			if (open && toastRef.current) {
+				// Toast.Viewport is 'ol' element
+				const viewport = toastRef.current.closest('ol')
+				window.requestAnimationFrame(() => {
+					if (viewport) {
+						// in case Toast.Viewport is overflows, we need to wait until the new toast
+						// inserted into the DOM before we scroll it to view
+						viewport.lastElementChild?.scrollIntoView({
+							behavior: 'smooth',
+							block: 'end',
+						})
+						// move Close All Notification button to position
+						if (isAction && visible && toastRef.current) {
+							viewport.insertBefore(
+								toastRef.current as unknown as Node,
+								viewport.firstElementChild as unknown as Node,
+							)
+						}
+					}
+				})
+			}
+		}, 0)
+	}, [open, isAction, visible])
```

Comment:
Is this not built-in functionality offered by radix? I feel like this imperative DOM interaction should not have to be done by us.

Reply:
i've cleanup toast in https://github.com/epicweb-dev/kcdshop/pull/30/commits/2baa1be5ee3966471a89935ef2fc858bed39adee

I am using Toast.Viewport forward ref to scroll it to top when new toast added